### ISSUE TITLE
As a user, I want valid dates pools to account for rescaling period in order to ensure I get all expected, rescaled pairs in the pools

### ISSUE THREAD

**Issue by HankHerr-NOAA** [NOAA-OWP/wres#357]
This issue was reported in the context of new `time_pools` developed and UAT'd in #245. See [this comment](https://github.com/NOAA-OWP/wres/issues/245#issuecomment-2460605186). For the declaration,

```yml
time_pools:
  - lead_times:
      minimum: 1
      maximum: 5
      unit: days
  - reference_dates:
      minimum: 1995-03-17T00:00:00Z
      maximum: 1995-03-20T00:00:00Z
  - valid_dates:
      minimum: 1995-03-18T00:00:00Z
      maximum: 1995-03-21T00:00:00Z
```

focusing on `valid_dates` and noting that the data is 6-hour but rescaled to a 24-hour mean, I expected values for valid date/times of 3/18, 3/19, and 3/20 6Z. However, I did not get a value for 3/18 6Z.

The guess is that, when gathering data for valid date pools, we are not accounting for the `period` of the `time_scale` as we do for other pools. I'll share the declaration that reproduced the problem noted here using `time_pools` in a comment, below. However, I still need to determine if it also handles top level `valid_dates` pools. 

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
This declaration produced the issue using `time_pools` is below. Tomorrow, I'll see if I can reproduce the problem using the top level pool and try to shrink it somewhat to make it easier to reproduce.

Thanks,

Hank

==========

```yml
label: BV Streamflow
observed:
  label: OBS Streamflow
  sources: /home/ISED/wres/wresTestData/issue92087/inputs/ABRN1_QME.xml
  variable: QME
  feature_authority: nws lid
  type: observations
  time_scale:
    function: mean
    period: 24
    unit: hours
predicted:
  label: HEFS Streamflow
  sources: /home/ISED/wres/wresTestData/issue92087/inputs/ABRN1_Streamflow_HEFS_Baseline_Validation.tgz
  variable: SQIN
  type: ensemble forecasts
  time_scale:
    function: mean
    period: 60
    unit: seconds
baseline:
  label: ESP Streamflow
  sources: /home/ISED/wres/wresTestData/issue92087/inputs/ABRN1_Streamflow_ESP_Baseline_Validation.tgz
  variable: SQIN
  type: ensemble forecasts
  time_scale:
    function: mean
    period: 60
    unit: seconds
features:
  - {observed: ABRN1, predicted: ABRN1, baseline: ABRN1}
unit: ft3/s
lead_times:
  minimum: 18
  maximum: 720
  unit: hours
lead_time_pools:
  period: 24
  frequency: 24
  unit: hours
# TESTING GitHub 245
# Testing #245
time_pools:
  - lead_times:
      minimum: 1
      maximum: 5
      unit: days
  - reference_dates:
      minimum: 1995-03-17T00:00:00Z
      maximum: 1995-03-20T00:00:00Z
  - valid_dates:
      minimum: 1995-03-18T00:00:00Z
      maximum: 1995-03-21T00:00:00Z 

time_scale:
  function: mean
  period: 24
  unit: hours
pair_frequency:
  period: 24
  unit: hours
cross_pair: exact
probability_thresholds: [0.01, 0.1, 0.5, 0.9, 0.95, 0.99, 0.995, 0.999]
season:
  minimum_day: 1
  minimum_month: 1
  maximum_day: 31
  maximum_month: 12
metrics:
  - mean error
  - continuous ranked probability skill score
  - box plot of errors by observed value
  - relative operating characteristic score
  - reliability diagram
  - brier skill score
  - pearson correlation coefficient
  - sample size
  - root mean square error
  - mean absolute error
  - relative operating characteristic diagram
  - bias fraction
  - brier score
  - quantile quantile diagram
  - box plot of errors by forecast value
  - rank histogram
  - continuous ranked probability score
duration_format: hours
decimal_format: '#0.000'
output_formats:
  - csv2
  - format: png
    orientation: lead threshold
  - pairs

**Comment by james-d-brown** [comment]
Sounds good. Another factor here is in-memory vs. database because this behavior is currently performed at the retriever (implementation) level. Perhaps it can be abstracted out, though.

**Comment by james-d-brown** [comment]
Looks like I don't have this dataset locally and it also isn't in the gdrive directory for large wres datasets (there is a directory for issue 92087, but it only includes temperature). Could you upload there, Hank?

**Comment by james-d-brown** [comment]
That said, it will help to corner this behavior with a unit test in any case.

**Comment by HankHerr-NOAA** [comment]
James,

The data supporting the declaration I already shared is currently being uploaded here in Google Drive:

https://drive.google.com/drive/folders/1V-Ex-1hcljTRB9yam8hDKigvHzHB-98L

They are pretty big files, so it may take a few minute for them to be uploaded. You should see three files eventually.

If you would like for me to reduce the size of the example, let me know and I'll do it tomorrow.

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
Actually, it was faster than I thought. Data is already there.

Hank

**Comment by james-d-brown** [comment]
Got it, thanks.

**Comment by james-d-brown** [comment]
Been busy with HEFS evaluations, but now back to this one.

**Comment by james-d-brown** [comment]
Have a failing unit test for retrieval of observations and forecasts from a database (since adjustment for the time-scale `period` only happens in that context). Making them pass...

**Comment by james-d-brown** [comment]
Have the unit tests passing so, in principle, should be fixed, but need to download the supplied data and test the evaluation reported in the OP.

**Comment by james-d-brown** [comment]
Putting aside whether there should be pairs for 6Z on 3/18 for a moment, these time windows look very odd:

```TimeWindowOuter[earliestReferenceTime=1995-03-17T00:00:00Z,latestReferenceTime=1995-03-20T00:00:00Z,earliestValidTime=1995-03-17T00:00:00Z,latestValidTime=1995-03-21T00:00:00Z,earliestLeadDuration=PT-2562047788015215H-30M-8S,latestLeadDuration=PT2562047788015215H30M7.999999999S], 
TimeWindowOuter[earliestReferenceTime=1995-03-17T00:00:00Z,latestReferenceTime=1995-03-20T00:00:00Z,earliestValidTime=1995-03-17T00:00:00Z,latestValidTime=1995-03-21T00:00:00Z,earliestLeadDuration=PT24H,latestLeadDuration=PT120H], 
TimeWindowOuter[earliestReferenceTime=1995-03-17T00:00:00Z,latestReferenceTime=1995-03-20T00:00:00Z,earliestValidTime=1995-03-18T00:00:00Z,latestValidTime=1995-03-21T00:00:00Z,earliestLeadDuration=PT-2562047788015215H-30M-8S,latestLeadDuration=PT2562047788015215H30M7.999999999S] ].
```

Struggling to understand these...

**Comment by james-d-brown** [comment]
Oh, wait a second! I copy/pasted this from the OP:

```
  - lead_times:
      minimum: 1
      maximum: 5
      unit: days
  - reference_dates:
      minimum: 1995-03-17T00:00:00Z
      maximum: 1995-03-20T00:00:00Z
  - valid_dates:
      minimum: 1995-03-18T00:00:00Z
      maximum: 1995-03-21T00:00:00Z
```

I don't think the above was intended at all. I think the following was intended, which means something completely different:

```
  - lead_times:
      minimum: 1
      maximum: 5
      unit: days
    reference_dates:
      minimum: 1995-03-17T00:00:00Z
      maximum: 1995-03-20T00:00:00Z
    valid_dates:
      minimum: 1995-03-18T00:00:00Z
      maximum: 1995-03-21T00:00:00Z
```

The former declares three separate pools, each one bounded by only one of the three dimensions, the others being unbounded (or, rather, bounded by whatever default constraints are used on `valid_dates`, `reference_dates` and `lead_times`). The latter contains a single pool that is explicitly bounded on all three dimensions.

In other words, you need to be a little careful when declaring pools explicitly because the list-like yaml declaration will mean something very different when you start a new list item instead of continuing the existing item. Recall that a pool can be declared with any combination of the three dimensions.

**Hank**, does that make sense?

**Comment by HankHerr-NOAA** [comment]
James,

I'm not saying the test makes any sense, but I was intending exactly what I shared in the description just to see what happens. That is, three independent pools. My general idea was to just throw something at it and then determine if I could make sense of the results. I think I actually performed your combined test earlier in the process, iirc; I just wanted to see how it would handle the three pools test afterwards. 

In this case, I largely saw what I expected in the pairs, except for the 6Z values not being there as expected in the OP.

Hank

**Comment by james-d-brown** [comment]
OK, gotcha, that's fine. However, I'm not certain the 03/18 6Z value should exist without trawling through the data more carefully. It doesn't exist, even if the pool dimensions are expanded a lot to include several days prior. You may want to double-check that. Regardless, I think the underlying issue is fixed with not expanding out the valid times to account for the scale period.

**Comment by james-d-brown** [comment]
The latter is now unit tested.

**Comment by HankHerr-NOAA** [comment]
Okay. Thanks. I'll have to double check what to expect, then. When I have time, of course.

Hank

**Comment by james-d-brown** [comment]
Creating some smaller evaluations based on system tests to double-check that it's working correctly. Unit tests say it is, so I expect these tests to confirm.

**Comment by HankHerr-NOAA** [comment]
Per your comment earlier, I'm now taking a look at the ABRN1 data to understand why the pools, 

```yaml
  - valid_dates:
      minimum: 1995-03-18T00:00:00Z
      maximum: 1995-03-21T00:00:00Z 
```

will not yield pairs for 3/18 at 6Z. I assumes its something in the declaration that I'm simply overlooking. 

Hank

**Comment by HankHerr-NOAA** [comment]
The observations are complete for March, 1995, with the following observations for 3/18 at 6Z:

```xml
<event date="1995-03-18" time="06:00:00" value="294" flag="0"/>
```

The forecasts are 12Z. So the 3/16 12Z forecast, for example, will allow for rescaled value with valid date 3/18 6Z (lead time 42 hours). 

I'm rerunning it now in order to take a fresh look at the pairs. I'll continue my investigation after a meeting and lunch.

Hank

**Comment by james-d-brown** [comment]
Similarly, this, based on system test scenario505:

```
observed:
  sources:
    - 25510317T00_FAKE2_observations.xml
  variable: DISCHARGE
predicted:
  sources:
    - 25510318T00_FAKE2_forecast.xml
  variable: STREAMFLOW
features:
  - {observed: FAKE2, predicted: FAKE2}
unit: CMS
time_pools:
  - valid_dates:
      minimum: 2551-03-18T05:00:00Z
      maximum: 2551-03-19T09:00:00Z
time_scale:
  function: mean
  period: 6
  unit: hours
metrics:
  - sample size
  - mean error
decimal_format: '#0.000000'
duration_format: hours
outp:ut_formats:
  - csv
  - pairs
  - csv2
```

With this, `25510317T00_FAKE2_observations.xml`:

```
<?xml version="1.0" encoding="UTF-8"?>
<TimeSeries xmlns="http://www.wldelft.nl/fews/PI" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.wldelft.nl/fews/PI http://fews.wldelft.nl/schemas/version1.0/pi-schemas/pi_timeseries.xsd" version="1.2">
    <timeZone>0.0</timeZone>
    <series>
        <header>
            <type>accumulative</type>
            <locationId>FAKE2</locationId>
            <parameterId>DISCHARGE</parameterId>
            <timeStep unit="second" multiplier="3600"/>
            <startDate date="2551-03-17" time="00:00:00"/>
            <endDate date="2551-03-20" time="12:00:00"/>
            <missVal>-999.0</missVal>
            <stationName>SOME FAKE PLACE 2</stationName>
            <lat>45.0</lat>
            <lon>105.0</lon>
            <units>CMS</units>
        </header>
        <event date="2551-03-17" time="00:00:00" value="313" flag="0"/>
        <event date="2551-03-17" time="01:00:00" value="317" flag="0"/>
        <event date="2551-03-17" time="02:00:00" value="331" flag="0"/>
        <event date="2551-03-17" time="03:00:00" value="347" flag="0"/>
        <event date="2551-03-17" time="04:00:00" value="349" flag="0"/>
        <event date="2551-03-17" time="05:00:00" value="353" flag="0"/>
        <event date="2551-03-17" time="06:00:00" value="359" flag="0"/>
        <event date="2551-03-17" time="07:00:00" value="367" flag="0"/>
        <event date="2551-03-17" time="08:00:00" value="373" flag="0"/>
        <event date="2551-03-17" time="09:00:00" value="379" flag="0"/>
        <event date="2551-03-17" time="10:00:00" value="383" flag="0"/>
        <event date="2551-03-17" time="11:00:00" value="389" flag="0"/>
        <event date="2551-03-17" time="12:00:00" value="397" flag="0"/>
        <event date="2551-03-17" time="13:00:00" value="401" flag="0"/>
        <event date="2551-03-17" time="14:00:00" value="409" flag="0"/>
        <event date="2551-03-17" time="15:00:00" value="419" flag="0"/>
        <event date="2551-03-17" time="16:00:00" value="421" flag="0"/>
        <event date="2551-03-17" time="17:00:00" value="431" flag="0"/>
        <event date="2551-03-17" time="18:00:00" value="433" flag="0"/>
        <event date="2551-03-17" time="19:00:00" value="439" flag="0"/>
        <event date="2551-03-17" time="20:00:00" value="443" flag="0"/>
        <event date="2551-03-17" time="21:00:00" value="449" flag="0"/>
        <event date="2551-03-17" time="22:00:00" value="457" flag="0"/>
        <event date="2551-03-17" time="23:00:00" value="461" flag="0"/>
        <event date="2551-03-18" time="00:00:00" value="463" flag="0"/>
        <event date="2551-03-18" time="01:00:00" value="467" flag="0"/>
        <event date="2551-03-18" time="02:00:00" value="479" flag="0"/>
        <event date="2551-03-18" time="03:00:00" value="487" flag="0"/>
        <event date="2551-03-18" time="04:00:00" value="491" flag="0"/>
        <event date="2551-03-18" time="05:00:00" value="499" flag="0"/>
        <event date="2551-03-18" time="06:00:00" value="503" flag="0"/>
        <event date="2551-03-18" time="07:00:00" value="509" flag="0"/>
        <event date="2551-03-18" time="08:00:00" value="521" flag="0"/>
        <event date="2551-03-18" time="09:00:00" value="523" flag="0"/>
        <event date="2551-03-18" time="10:00:00" value="541" flag="0"/>
        <event date="2551-03-18" time="11:00:00" value="547" flag="0"/>
        <event date="2551-03-18" time="12:00:00" value="557" flag="0"/>
        <event date="2551-03-18" time="13:00:00" value="563" flag="0"/>
        <event date="2551-03-18" time="14:00:00" value="569" flag="0"/>
        <event date="2551-03-18" time="15:00:00" value="571" flag="0"/>
        <event date="2551-03-18" time="16:00:00" value="577" flag="0"/>
        <event date="2551-03-18" time="17:00:00" value="587" flag="0"/>
        <event date="2551-03-18" time="18:00:00" value="593" flag="0"/>
        <event date="2551-03-18" time="19:00:00" value="599" flag="0"/>
        <event date="2551-03-18" time="20:00:00" value="601" flag="0"/>
        <event date="2551-03-18" time="21:00:00" value="607" flag="0"/>
        <event date="2551-03-18" time="22:00:00" value="613" flag="0"/>
        <event date="2551-03-18" time="23:00:00" value="617" flag="0"/>
        <event date="2551-03-19" time="00:00:00" value="619" flag="0"/>
        <event date="2551-03-19" time="01:00:00" value="631" flag="0"/>
        <event date="2551-03-19" time="02:00:00" value="641" flag="0"/>
        <event date="2551-03-19" time="03:00:00" value="643" flag="0"/>
        <event date="2551-03-19" time="04:00:00" value="647" flag="0"/>
        <event date="2551-03-19" time="05:00:00" value="653" flag="0"/>
        <event date="2551-03-19" time="06:00:00" value="659" flag="0"/>
        <event date="2551-03-19" time="07:00:00" value="661" flag="0"/>
        <event date="2551-03-19" time="08:00:00" value="673" flag="0"/>
        <event date="2551-03-19" time="09:00:00" value="677" flag="0"/>
        <event date="2551-03-19" time="10:00:00" value="683" flag="0"/>
        <event date="2551-03-19" time="11:00:00" value="691" flag="0"/>
        <event date="2551-03-19" time="12:00:00" value="701" flag="0"/>
        <event date="2551-03-19" time="13:00:00" value="709" flag="0"/>
        <event date="2551-03-19" time="14:00:00" value="719" flag="0"/>
        <event date="2551-03-19" time="15:00:00" value="727" flag="0"/>
        <event date="2551-03-19" time="16:00:00" value="733" flag="0"/>
        <event date="2551-03-19" time="17:00:00" value="739" flag="0"/>
        <event date="2551-03-19" time="18:00:00" value="743" flag="0"/>
        <event date="2551-03-19" time="19:00:00" value="751" flag="0"/>
        <event date="2551-03-19" time="20:00:00" value="757" flag="0"/>
        <event date="2551-03-19" time="21:00:00" value="761" flag="0"/>
        <event date="2551-03-19" time="22:00:00" value="769" flag="0"/>
        <event date="2551-03-19" time="23:00:00" value="773" flag="0"/>
        <event date="2551-03-20" time="00:00:00" value="787" flag="0"/>
        <event date="2551-03-20" time="01:00:00" value="797" flag="0"/>
        <event date="2551-03-20" time="02:00:00" value="809" flag="0"/>
        <event date="2551-03-20" time="03:00:00" value="811" flag="0"/>
        <event date="2551-03-20" time="04:00:00" value="821" flag="0"/>
        <event date="2551-03-20" time="05:00:00" value="823" flag="0"/>
        <event date="2551-03-20" time="06:00:00" value="827" flag="0"/>
        <event date="2551-03-20" time="07:00:00" value="829" flag="0"/>
        <event date="2551-03-20" time="08:00:00" value="839" flag="0"/>
        <event date="2551-03-20" time="09:00:00" value="853" flag="0"/>
        <event date="2551-03-20" time="10:00:00" value="857" flag="0"/>
        <event date="2551-03-20" time="11:00:00" value="859" flag="0"/>
        <event date="2551-03-20" time="12:00:00" value="863" flag="0"/>
    </series>
</TimeSeries>

```

And this, `25510318T00_FAKE2_forecast.xml`:

```
<?xml version="1.0" encoding="UTF-8"?>
<TimeSeries xmlns="http://www.wldelft.nl/fews/PI" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.wldelft.nl/fews/PI http://fews.wldelft.nl/schemas/version1.0/pi-schemas/pi_timeseries.xsd" version="1.5">
    <timeZone>0.0</timeZone>
    <series>
        <header>
            <type>accumulative</type>
            <locationId>FAKE2</locationId>
            <parameterId>STREAMFLOW</parameterId>
            <timeStep unit="second" multiplier="10800"/>
            <startDate date="2551-03-18" time="03:00:00"/>
            <endDate date="2551-03-19" time="12:00:00"/>
            <forecastDate date="2551-03-18" time="00:00:00"/>
            <missVal>-999</missVal>
            <stationName>SOME FAKE PLACE 2</stationName>
            <units>CMS</units>
            <creationDate>2018-01-11</creationDate>
            <creationTime>20:45:50</creationTime>
        </header>
        <event date="2551-03-18" time="03:00:00" value="131" flag="0"/>
        <event date="2551-03-18" time="06:00:00" value="137" flag="0"/>
        <event date="2551-03-18" time="09:00:00" value="139" flag="0"/>
        <event date="2551-03-18" time="12:00:00" value="149" flag="0"/>
        <event date="2551-03-18" time="15:00:00" value="151" flag="0"/>
        <event date="2551-03-18" time="18:00:00" value="157" flag="0"/>
        <event date="2551-03-18" time="21:00:00" value="163" flag="0"/>
        <event date="2551-03-19" time="00:00:00" value="167" flag="0"/>
        <event date="2551-03-19" time="03:00:00" value="173" flag="0"/>
        <event date="2551-03-19" time="06:00:00" value="179" flag="0"/>
        <event date="2551-03-19" time="09:00:00" value="181" flag="0"/>
    </series>
</TimeSeries>

```

Produces this, which is expected:

```
FEATURE NAME,FEATURE GROUP NAME,VARIABLE NAME,EARLIEST ISSUE TIME,LATEST ISSUE TIME,EARLIEST VALID TIME,LATEST VALID TIME,EARLIEST LEAD TIME [MEAN OVER PAST PT6H],LATEST LEAD TIME [MEAN OVER PAST PT6H],REFERENCE TIME,VALID TIME,LEAD DURATION [MEAN OVER PAST PT6H],OBSERVED IN CMS,PREDICTED IN CMS
FAKE2,FAKE2-FAKE2,DISCHARGE,-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,2551-03-18T05:00:00Z,2551-03-19T09:00:00Z,PT-2562047788015215H-30M-8S,PT2562047788015215H30M7.999999999S,2551-03-18T00:00:00Z,2551-03-18T06:00:00Z,PT6H,487.666667,134.000000
FAKE2,FAKE2-FAKE2,DISCHARGE,-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,2551-03-18T05:00:00Z,2551-03-19T09:00:00Z,PT-2562047788015215H-30M-8S,PT2562047788015215H30M7.999999999S,2551-03-18T00:00:00Z,2551-03-18T12:00:00Z,PT12H,533.000000,144.000000
FAKE2,FAKE2-FAKE2,DISCHARGE,-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,2551-03-18T05:00:00Z,2551-03-19T09:00:00Z,PT-2562047788015215H-30M-8S,PT2562047788015215H30M7.999999999S,2551-03-18T00:00:00Z,2551-03-18T18:00:00Z,PT18H,576.666667,154.000000
FAKE2,FAKE2-FAKE2,DISCHARGE,-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,2551-03-18T05:00:00Z,2551-03-19T09:00:00Z,PT-2562047788015215H-30M-8S,PT2562047788015215H30M7.999999999S,2551-03-18T00:00:00Z,2551-03-19T00:00:00Z,PT24H,609.333333,165.000000
FAKE2,FAKE2-FAKE2,DISCHARGE,-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,2551-03-18T05:00:00Z,2551-03-19T09:00:00Z,PT-2562047788015215H-30M-8S,PT2562047788015215H30M7.999999999S,2551-03-18T00:00:00Z,2551-03-19T06:00:00Z,PT30H,645.666667,176.000000
```

Specifically, it includes the pair at 6Z on 03/18, despite half of the data falling outside of the valid datetime window.

Neither of these two tests produce the expected pairs with v6.27.

**Comment by james-d-brown** [comment]
> Per your comment earlier, I'm now taking a look at the ABRN1 data to understand why the pools,
> 
> ```yaml
>   - valid_dates:
>       minimum: 1995-03-18T00:00:00Z
>       maximum: 1995-03-21T00:00:00Z 
> ```
> 
> will not yield pairs for 3/18 at 6Z. I assumes its something in the declaration that I'm simply overlooking.
> 
> Hank

Oh, right, you will need to recall our earlier discussion about the meaning of `minimum` in the context of explicit pools and what we decided, namely to make the application minimum exclusive, event if that is slightly counterintuitive. The reason being that, when a user declares a pool sequence, it becomes more intuitive to declare the lower boundary of one pool as the upper boundary of the next. Try `minimum: 1995-03-17T00:00:00Z`, for example.

**Comment by james-d-brown** [comment]
Oh, wait, the `minimum` is 0Z, not 6Z, carry on.

**Comment by HankHerr-NOAA** [comment]
At this point, I think I need to wait for the fix for this ticket to be pushed to master before investigating further. I've confirmed that the behavior I'm seeing when `valid_dates` is at the top level (i.e., not in `time_pools`) is identical to when its declared in `time_pools` as far as only including pairs for 3/19 and 3/20 6Z.  I believe you said that, even with your change, you do not see 3/18 6Z pairs. To understand why, I may need to see it in action, because, again, there are clearly forecasts available that will provide a forecast value for 3/18 6Z (e.g., the 3/16 12Z forecast) and the observed value is not missing.

Note to self: I have the 3 declarations to use (full declaration, one with only the `valid_dates` time pool, and one with `valid_dates` at the top level) are all in the `github245` directory on the ISED dev machine (since this originally came up during UAT of #245). Just run them again and compare the resulting pairs. 

I'll revisit this once the code is in master,

Hank

**Comment by james-d-brown** [comment]
I don't see it, although I did try to reduce the size of the evaluation a lot, so perhaps I made a mistake. Yeah, I can push and merge. I'm pretty confident it is now working as expected, though, because it is unit tested and I've run several end-to-end ad-hoc evaluations to confirm it.

**Comment by HankHerr-NOAA** [comment]
With the code pushed, I've just pulled it and will run the same evaluation I've been using to test and see if I can get data for 3/18 6Z.  

Hank

**Comment by HankHerr-NOAA** [comment]
If I remove `lead_times` and `lead_times_pools` from the declaration, I still only see pairs for 3/19 and 3/20 6Z.

Hank

**Comment by HankHerr-NOAA** [comment]
I also see only pairs for 3/19 and 3/20 6Z if I remove the `baseline`. So without `baseline`, `lead_times`, and `lead_time_pools`, I still do not see any pairs for 3/18 6Z even though it is within the defined `valid_dates`. 

I've also confirmed the revision: `20241121-e85bfa5-dev`. 

Thinking out loud... I've defined valid dates pools that start on 3/18 0Z. The scale is 24-hours. The observed values are 24-hour means computed at 6Z of every day, so the rescaled forecasts are computed to end at 6Z of every day.  A 3/18 6Z forecast value would require data for 3/17 12Z, 18Z and 3/18 0Z, 6Z.  

As shown below, the 3/16/1995 12Z forecast includes the needed values, and the rescaled value would have a lead time of PT42H. 

I'm at a bit of a loss as to why I don't see the expected pair. Am I missing something?

Hank  

==========

```xml
    <series>
        <header>
            <type>instantaneous</type>
            <locationId>ABRN1</locationId>
            <parameterId>SQIN</parameterId>
            <ensembleId>MEFP</ensembleId>
            <ensembleMemberIndex>1961</ensembleMemberIndex>
            <timeStep unit="second" multiplier="21600"/>
            <startDate date="1995-03-16" time="18:00:00"/>
            <endDate date="1995-04-15" time="12:00:00"/>
            <forecastDate date="1995-03-16" time="12:00:00"/>
            <missVal>-999.0</missVal>
            <stationName>ABRN1: AUBURN NE 1NE</stationName>
            <units>CMS</units>
            <creationDate>2019-03-01</creationDate>
            <creationTime>07:41:19</creationTime>
        </header>
        <event date="1995-03-16" time="18:00:00" value="26.28568" flag="0"/>
        <event date="1995-03-17" time="00:00:00" value="21.91186" flag="0"/>
        <event date="1995-03-17" time="06:00:00" value="17.960527" flag="0"/>
        <event date="1995-03-17" time="12:00:00" value="14.9450655" flag="0"/>
        <event date="1995-03-17" time="18:00:00" value="12.41184" flag="0"/>
        <event date="1995-03-18" time="00:00:00" value="10.431927" flag="0"/>
        <event date="1995-03-18" time="06:00:00" value="8.665522" flag="0"/>
        <event date="1995-03-18" time="12:00:00" value="7.5124598" flag="0"/>
```

**Comment by HankHerr-NOAA** [comment]
I've confirmed that the 3/16 12Z forecast is seen. The two pairs have lead times PT66H and PT90H. Here are the pairs:

```
ABRN1,ABRN1-ABRN1,QME,-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,1995-03-18T00:00:00Z,1995-03-21T00:00:00Z,PT-2562047788015215H-30M-8S,PT2562047788015215H30M7.999999999S,1995-03-16T12:00:00Z,1995-03-19T06:00:00Z,PT66H,234.000,948.148,225.678,902.085,225.678,225.678,225.678,225.678,225.678,225.678,228.118,225.678,546.410,225.678,225.678,225.678,225.678,225.678,225.678,3951.318,225.678,225.678,229.995,225.678,235.285,225.678,974.748,2420.683,254.233,225.678,225.678,225.678,2444.315,761.593,225.678,375.230,225.678,225.678,675.865,225.678,892.803,225.678,225.678,688.533,225.678,225.678,827.125,225.678,225.678,229.450,225.678
ABRN1,ABRN1-ABRN1,QME,-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,1995-03-18T00:00:00Z,1995-03-21T00:00:00Z,PT-2562047788015215H-30M-8S,PT2562047788015215H30M7.999999999S,1995-03-16T12:00:00Z,1995-03-20T06:00:00Z,PT90H,212.000,784.523,163.595,1074.958,163.645,163.645,163.645,163.595,163.595,163.645,172.225,163.645,568.443,163.595,163.645,163.595,163.595,163.645,163.645,3239.168,163.635,163.595,357.148,163.595,337.013,163.635,1620.615,2252.910,199.585,163.645,163.595,163.635,2945.830,621.313,163.595,272.278,163.645,163.645,680.155,163.635,1225.813,163.595,163.595,1008.625,163.645,163.645,922.000,163.595,163.595,168.293,163.595
```

**Comment by HankHerr-NOAA** [comment]
Removing `pair_frequency` did not change the results. Still no pairs for 3/18 6Z.

Hank

**Comment by HankHerr-NOAA** [comment]
James:

I reduced the evaluation I'm using to test to what I have provided below using the ABRN1 data shared earlier. It's still not producing pairs for 3/18 6Z, even though, as I described above, I at least expect something from the forecast at 3/16 12Z with a 42h lead time.  When I run it in debug mode, I encountered this log message:

> 2024-11-25T14:17:08.545+0000  [Pool Thread 1] DEBUG RescaledTimeSeriesPlusValidation - While rescaling time-series with metadata TimeSeriesMetadata[timeScale=[INSTANTANEOUS],referenceTimes={T0=1995-03-16T12:00:00Z},variableName=SQIN,feature=Feature[name=ABRN1,description=ABRN1: AUBURN NE 1NE,srid=0,wkt=],unit=CMS], encountered 1 detailed validation warnings, as follows:
    EvaluationStatusMessage[LEVEL=DEBUG,STAGE=RESCALING,MESSAGE=While attempting to upscale a collection of 1 events to an evaluation time scale of [PT24H,MEAN] ending at 1995-03-18T06:00:00Z, discovered fewer than two events in the collection, which cannot produce an upscaled value.]

Its specifically referring to the 3/16 12Z forecast, stating that it attempts to upscale a value for 3/18 6Z, but discovered only one event. I believe that is the problem this ticket is supposed to resolve, correct? The forecast definitely has the needed events, and I think this ticket should result in it looking back beyond the `valid_dates` `minimum` in order to account for the `time_scale` `period`. That should allow it to find more than one.

Let me know what you think.

Thanks,

Hank

===========================

```yaml
label: BV Streamflow
observed:
  label: OBS Streamflow
  sources: /home/ISED/wres/wresTestData/issue92087/inputs/ABRN1_QME.xml
  variable: QME
  feature_authority: nws lid
  type: observations
predicted:
  label: HEFS Streamflow
  sources: /home/ISED/wres/wresTestData/issue92087/inputs/ABRN1_Streamflow_HEFS_Baseline_Validation.tgz
  variable: SQIN
  type: ensemble forecasts
features:
  - {observed: ABRN1, predicted: ABRN1} 
unit: ft3/s
valid_dates:
  minimum: 1995-03-18T00:00:00Z
  maximum: 1995-03-21T00:00:00Z 

time_scale:
  function: mean
  period: 24
  unit: hours

metrics:
  - mean error
  - pearson correlation coefficient
  - sample size
  - root mean square error
  - mean absolute error
output_formats:
  - csv2
  - format: png
    orientation: lead threshold
  - pairs
```

**Comment by james-d-brown** [comment]
Correct, but that debug log message simply indicates that insufficient forecast data was found for upscaling, it doesn't indicate what data was requested. For that, you need to inspect the retrieval script, which is also available with debug looking.

Anyway, let me take a look to see whether there's something here that hasn't been cornered with a unit test...

**Comment by james-d-brown** [comment]
Right, see my first post in this ticket. However, I thought all of your testing was against a db. FWIW, you should pretty much always test against a db when performing UAT or at least do both, because a db is our default cowres deployment and there are a bunch of different pathways for in-memory. From my POV, this whole ticket has so far been about retrieval from a db because that is the only scenario in which a subset of data is requested. With in-memory retrieval, the full time-series are always in-memory, by definition, so there is no need for specially filtered retrieval scripts for a subset of data - in fact, retrieval is just a facade with the in-memory pathway.

**Comment by james-d-brown** [comment]
The only commits to this ticket have focused on db-specific implementations. If there is a separate issue for in-memory evaluations with valid-time constraints, that is a separate ticket, I think.

**Comment by HankHerr-NOAA** [comment]
The Postgres database version gave me the pairs I expected.  

I'll create that ticket now,

Hank

**Comment by HankHerr-NOAA** [comment]
Created #369. I think this can be closed. 

Hank

**Comment by HankHerr-NOAA** [comment]
Actually, the status says UAT already. 

The test I just ran does, in fact, UAT this ticket, since I saw the pairs I expected. However, I'm running the original declaration with a database just to be sure that I can find what I expect to find in the outputs,

Hank

**Comment by james-d-brown** [comment]
Thanks, Hank. Closing, but retaining the UAT status for deployment time.

**Comment by HankHerr-NOAA** [comment]
Definitely easier to scan the output for the reduced version of the declaration. For the full, I did find the pairs by first searching for "3-21T00" for the pool of interest, and then "3-18T06" to find the pairs that were missing before.  The impact is not easy to judge in the evaluation graphics.  

Anyway, I'm seeing what I expect in the database run. We can run the test again during deployment time,

Hank
